### PR TITLE
Usage integration

### PR DESCRIPTION
**Pull request by akley**
Hi Stéphane & Pierre,

I've finished the client improvement to use server side Usage-API. First step was to impelment the usage data creation.

Cheers,
Alex

### CODE REVIEW

File: src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package org.killbill.billing.client.model;
```
Comment (pierre) [inline src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java]: Note: need to add copyright header for RAT.

File: src/main/java/org/killbill/billing/client/model/UnitUsage.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
+package org.killbill.billing.client.model;
```
Comment (pierre) [inline src/main/java/org/killbill/billing/client/model/UnitUsage.java]: Note: need to add copyright header for RAT.

File: src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.util.List;
+import java.util.UUID;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class SubscriptionUsage  {
```
Comment (pierre) [inline src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java]: Should we call it `SubscriptionUsageRecord` for consistency with `SubscriptionUsageRecordJson`?

File: src/main/java/org/killbill/billing/client/model/UnitUsage.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.util.List;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class UnitUsage {
```
Comment (pierre) [inline src/main/java/org/killbill/billing/client/model/UnitUsage.java]: Should we call it `UnitUsageRecord` for consistency with `UnitUsageRecordJson`?

File: src/main/java/org/killbill/billing/client/model/Usage.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
+package org.killbill.billing.client.model;
```
Comment (pierre) [inline src/main/java/org/killbill/billing/client/model/Usage.java]: Note: need to add copyright header for RAT.

File: src/main/java/org/killbill/billing/client/model/Usage.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+import org.joda.time.LocalDate;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class Usage {
```
Comment (pierre) [inline src/main/java/org/killbill/billing/client/model/Usage.java]: Should we call it `UsageRecord` for consistency with `UsageRecordJson`?

**Comment by pierre** [comment]
Looking good! See my comments. Also, could you add a corresponding [server](https://github.com/killbill/killbill/blob/master/profiles/killbill/src/test/java/org/killbill/billing/jaxrs) test?

File: src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java (hunk 1 above)
Comment (akley) [inline src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java]: Sorry Pierre I don't understand exactly what do you mean: for the files we must add copyright header? 
RAT - what does it mean?

File: src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java (hunk 3 above)
Comment (akley) [inline src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java]: No question - it's a good idea! But it deoends, my intention was to follow the naming of the other client model classes. We can rename the classes. 
If you want that, then it is better if I create new Pull Request?

**Comment by akley** [comment]
For the tests, I can try it, but I have only integrated the write method (create usage). So actually we can not retrieve usages from server via client.

File: src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java (hunk 1 above)
Comment (pierre) [inline src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java]: > for the files we must add copyright header?

Yes, something similar to the other files to license them under the Apache License 2.0.

> RAT - what does it mean?

It's our auditing tool (http://creadur.apache.org/rat/), which makes sure all files are properly licensed.

File: src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java (hunk 3 above)
Comment (pierre) [inline src/main/java/org/killbill/billing/client/model/SubscriptionUsage.java]: Let's follow the current naming convention (i.e. the Kill Bill model name minus the Json suffix). You can update the same PR, that's fine.

**Comment by pierre** [comment]
> For the tests, I can try it, but I have only integrated the write method (create usage). So actually we can not retrieve usage data from server via client.

Yup, but it shouldn't be too hard since the models now exist?

For reference: https://github.com/killbill/killbill/blob/master/jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/UsageResource.java

**Comment by akley** [comment]
For the tests, no problem, but I can complete this missing piece in some days, because we are currently in test abd release stage of our software.

**Comment by akley** [comment]
Is file renaming also possible in the same PR? Or it's better to do this in a new PR?

**Comment by pierre** [comment]
> For the tests, no problem, but I can only complete this missing piece in some days, because we are currently in test abd release stage of our software.

No worries, take your time.

> Is file renaming also possible in the same PR? Or it's better to do this in a new PR?

Yes, you can. The PR will mirror the changes in your `usage-integration` branch.

**Comment by akley** [comment]
Class changes (renaming, license and copyright) done.

**Comment by sbrossie** [comment]
Any update on the PR? Did you have a chance to add some tests?

**Comment by akley** [comment]
So my vacation is over ...
This week we figured out some use cases to integrate read functionality. In this context we also write some tests, so it is a good chance for the PR ;-)

**Comment by sbrossie** [comment]
@akley Will merge your PR (the tests need to be added in the [killbill repo](https://github.com/killbill/killbill/tree/master/profiles/killbill/src/test/java/org/killbill/billing/jaxrs) so that would be a different PR anyways).

**Comment by sbrossie** [comment]
I think this was closed by mistake....

@akley are you still working on that? Will you add some tests at some point in [killbill repo](https://github.com/killbill/killbill/tree/master/profiles/killbill/src/test/java/org/killbill/billing/jaxrs)?

**Comment by MusharrafSakran** [comment]
@akley @sbrossie @pierre any updates on this PR ? because we had a use case similar to the purpose for this PR.

**Comment by pierre** [comment]
I'll take a look.

**Comment by pierre** [comment]
Merged and released (v0.20).

**Pull request merged**